### ISSUE TITLE
Interactive Functionality with Drip

### ISSUE THREAD

**Issue by jrosebr1** [realbig/PL_PYIS_HELPSCOUT_DRIP#8]
Building on #7, I have a few more questions on what can be accomplished via HelpScout's interface. Is it possible to:

1. Have the ability to add/remove tags from the sidebar?
2. Display a modal window inside of HelpScout when a specific button is clicked in the sidebar? My goal here would be to select a "Workflow" to apply to the subscriber (useful if a customer purchases a copy of my book, but for whatever reason Drip isn't notified; a customer wants a refund and needs to be removed from follow-up campaigns, etc.)
3. What access do we have to the "activity feed" of a subscriber. Is it possible to access this via the API?

I can see _a lot_ of extensions to this plugin and I have a ton of ideas but I need to know what we can feasibly do within HelpScout's constraints.

**Comment by d4mation** [comment]
These all sound like super cool ideas, but as far as I can tell HelpScout doesn't support any back-and-forth communication. At least not natively.

It could potentially be possible for us to include a _ton_ of inline-JavaScript in the HTML response and then create our own little AJAX calls using that to hit some custom endpoints on the WP site to then communicate with Drip, but I have a feeling that would be very inefficient and buggy.

**Comment by jrosebr1** [comment]
I agree, that sounds very buggy and inefficient. Thanks for the added information. If you could elaborate on the current capabilities of HelpScout that would be super helpful as we continue to develop and release different versions of the software.

**Comment by d4mation** [comment]
HelpScout makes a request to the WP server and then the WP server can return HTML. In this case WP makes a request to Drip and uses that data to build an HTML response.

Unfortunately that's about as far as HelpScout Apps' capabilities extend :/

If we wanted to do anything cool like you've outlined above we'd need a way to load our own JavaScript within the HelpScout dashboard.

Here's HelpScout's documentation on it:

http://developer.helpscout.net/custom-apps/dynamic/

**Comment by d4mation** [comment]
It only makes that initial request to the WP server when a conversation is opened, so we cannot (natively within HelpScout) have it send requests when something is clicked for example.

**Comment by jrosebr1** [comment]
Thanks for the extra information, I appreciate it. So we can't do any type of back-and-forth communication from _within_ HelpScout. But if we provided a _link_ on the sidebar that opened a pop-up window (hard to believe I'm saying those words in 2017), and this pop-up window was hosted on the WP site, we could then edit tags, apply workflows, etc.?

**Comment by d4mation** [comment]
Potentially, but at that point I think I'd be looking into either including some JavaScript in the HTML Request or otherwise hosting a JavaScript file on the WP site for HelpScout to include via a `<script>` tag in the HTML Request.

Outside of AUTH concerns (The WP Rest API doesn't support that horribly well at the moment without using an external plugin), I think JavaScript may be the best option.

**Comment by jrosebr1** [comment]
Thanks Eric, I appreciate it. We should probably put a pin in this conversation for now, I'm just looking at what is possible. This might make for a good SaaS app if moved outside the WP ecosystem, but the features I mentioned above would have to be there.

**Comment by jrosebr1** [comment]
Since it seems like it would take a lot of time and effort to build out the code to allow us to interact with Drip *inside* of HelpScout, I think for the time being the sidebar should simply provide a link to the subscriber's page in Drip. At the very least this will allow me to click on this link and remove any tags/subscriptions that way.

Again, I would _really_ love to be able to edit tags and apply workflows from _inside_ HelpScout, but if that isn't possible, let's start with a simple link to the subscriber's profile page and go from there.

**Comment by d4mation** [comment]
I went ahead and added the link to the Subscriber into the Response. 
https://github.com/realbig/PL_PYIS_HELPSCOUT_DRIP/releases/tag/v0.4.1

I absolutely agree about adding this interactive functionality and think it would be really cool. I just want to ensure we handle that in the most efficient and secure way.

If we could load some JS within HelpScout (Either via Inline JS in the Response HTML or by including a `<script>` tag that loads it from WP), we should be able to handle this. Either way the JavaScript will needed to be loaded into HelpScout. But there's two problems I foresee security-wise:

1. If we have WP be the middle-man in this scenario as well, we will want to have some sort of Authentication set up for the WP REST API. We're getting around this now with the "Secret Key", but for any other interactions we no longer have access to this Secret Key. There is an official implementation of OAUTH for the REST API, but it isn't a part of WP Core yet. But then again, the REST API itself wasn't either until WP v4.4.
    - https://wordpress.org/plugins/rest-api-oauth1/
2. We could send your Drip Username and Password as plaintext in the Response HTML and use that to construct Drip API requests directly from HelpScout via JavaScript. Doing this gives me the heebie-jeebies though.

**Comment by jrosebr1** [comment]
I agree, sending the plaintext username + password in the response HTML is an absolute no.

That said, relying strictly on the WP REST API [also gives me concern](https://blog.sucuri.net/2017/02/wordpress-rest-api-vulnerability-abused-in-defacement-campaigns.html). That said, it seems much more reliable than sending plaintext passwords over the wire.

**Comment by jrosebr1** [comment]
Follow up question: Imagine we were to build this app _without_ WordPress and instead wanted to build a standalone web application. Would we still run into the same issues?

**Comment by d4mation** [comment]
That exploit only pertained to WP v4.7.0 and v4.7.1, every other release before then actually was immune to it. Unless your Staging site is running a different WP version than your Live, you're running v4.4.2 which while outdated isn't effected by it at all :+1:.

v4.7 included content management endpoints which they weren't careful enough with unfortunately. v4.4 just included the ability to create endpoints basically, which is entirely safe provided you're smart about it (which is what we're doing).

We could build a stand-alone Web App like a Chrome Extension that handles the Tag Management functionality. It wouldn't be any less secure as what we're doing now: it would just load on HelpScout conversation pages, add some DOM for little `x` buttons or something, then preform the necessary API interactions with Drip. All the current functionality could be handled by this current codebase itself.

That actually may be a nice feature in-and-of-itself, actually. Tags and other information would show on any device you access the conversation from, but you can only manage Tags if you've installed the correct extension and have the credentials entered into it.

**Comment by jrosebr1** [comment]
Thanks for the updated information, I appreciate it. As I mentioned earlier, I know a few people who utilize both Drip + HelpScout, so it's entirely possible that something like this could be spun off into a separate product. The issue would just be resolving how to communicate with HelpScout in a secure way without sending plaintext information.

So if I understand your suggestion properly, the best route would be to create a Chrome Extension to handles _interactive_ features with HelpScout. That makes sense, my concern would be why not have the entire "app" as a Chrome Extension then? Why bother with the server side code?

**Comment by d4mation** [comment]
The server-side code is primarily a "Quality-of-life" feature at this point.

The initial project specs didn't include any interactive functionality, so all it needed to do was show what information was _currently_ in Drip. 

This works great for cases like viewing the HelpScout conversation on your cell phone or on another computer where a browser extension (or even a specific browser) may not be installed. The data will always be there, no matter how you're viewing it.

It could all be handled by a browser extension, but then you'd lose that ability to always at least be able to _view_ the data from any device. 

That may not be an important bit of functionality, but it was a decision I made when reading the project specs that since all it had to do was _render_ data and not interact with it that HelpScout's "Pure HTML Response" Integration System would work perfectly considering future potential enhancements like adding other bits of data to it.

---

This being said, @brashrebel and I are talking with HelpScout to see if there's perhaps a "best practice" for handling this within their ecosystem. There may be some better way to do this outside of ramming JS in there either via the HTML Response or a Chrome Extension.

**Comment by jrosebr1** [comment]
Got it, that makes sense. Let me know what HelpScout says regarding the best practice!
